### ISSUE TITLE
systemd-resolved forbids rrsig type query

### ISSUE THREAD

**Issue by pemensik** [systemd/systemd#17223]
**systemd version the issue has been seen with**
> ...systemd-246.6-2.fc34.x86_64

**Used distribution**
 > Fedora 34 Rawhide

**Expected behaviour you didn't see**
 > It just delivers any response from forwarders or answers with NOERROR and empty answer section.

**Unexpected behaviour you saw**
 > resolvectl -t rrsig query localhost
> localhost: resolve call failed: Specified resource record type 46 may not be used in a query.

**Steps to reproduce the problem**
dig -t RRSIG @127.0.0.53 localhost

DNSSEC supported: yes

**Comment by poettering** [comment]
Hmm? asking for just an rrsig doesn#t really make much sense. we don't allow that in resolvectl, because why should we? what is the problem precisely?

**Comment by pemensik** [comment]
it should allow any valid query. if you want to know whether some domain is signed, you should be able to fetch only signatures. If that record exists, resolved has to handle it. It is just as good as A, AAAA or URI type. Query of DS is not the same. If no such record exist, return NOERROR status with no answer in body.

I know you don't allow it in resolvectl and you don't allow it via DNS query. That's a bug. Why do you need to verify what query types I want? Can you point me to any other implementation, which does not allow such query?

Is there any list of other query types it would not allow me to send?

**Comment by poettering** [comment]
See https://github.com/systemd/systemd/blob/master/src/resolve/dns-type.c#L66

Also I remember while testing resolved we found some wifi routers that rebooted if you sent them RRSIG requests, that's why we refuse them too. 

Let's not forget that resolved is a tightly sandboxed DNS/mDNS/LLMNR/… stack that is supposed to be useful as a gatekeeper for name resolution requests: a client that itself is sandboxed should be able to use it to do reasonably safe lookups only, but not be able to send arbitrarily formatted packets. Now one could argue that RRSIG queries are not yet "arbitrarily formatted", but I think we should still filter them out.

**Comment by pemensik** [comment]
Okay, OPT, TSIG and TKEY are correct. RRSIG is a bit special, but it should work nonetheless.

If some crappy routers cannot handle rrsig query, it should be reported to their manufacturer. Do not fix bugs of others in your software. If you want to be kind to such people, provide them configurable blacklist to selected query types. But do not block it for everyone because some obscure devices.

It should allow sending packets it does not completely understand, in order to work well with future changes of protocol. If RRSIG works as packet of death, blocking it in local resolver does not really help. Only firewall would protect it. Or better, software update from vendor, because this is clearly security issue you cannot fix inside resolved.

Since such queries are not usually done by common software, they should be forwarded as requested. User should get reply whatever he asked for. It should filter only what does not belong there, which is not case for rrsig.

**Comment by mercora** [comment]
i am not certain if it is related but queries for DS records are failing for delv (a "DNS lookup and validation utility")  too, but not via FormErr but ServFail which makes delv report the trust chain is broken.

```
$ delv lileth.net
;; no valid RRSIG resolving 'net/DS/IN': 127.0.0.53#53
;; broken trust chain resolving 'lileth.net/A/IN': 127.0.0.53#53
;; resolution failed: broken trust chain
```
```
$ delv net.
;; resolution failed: SERVFAIL
```
```
$ delv lileth.net @172.18.11.1
;; resolution failed: ncache nxrrset
; negative response, fully validated
; lileth.net.           1332    IN      \-A     ;-$NXRRSET
; lileth.net. SOA ns1.lileth.net. mercora.lileth.net. 2020081853 14400 3600 3600000 3600
; lileth.net. RRSIG SOA ...
; 0BGHOEC1TRN4MKS9Q2RG3OV1J2976QML.lileth.net. NSEC3 1 0 10 2426CEF4B9FB3531 0E5PH4CVGG1N9SLBT9VFNV1E0S91K396 NS SOA MX TXT RRSIG DNSKEY NSEC3PARAM
; 0BGHOEC1TRN4MKS9Q2RG3OV1J2976QML.lileth.net. RRSIG NSEC3 ...
```

However, doing the query via resolvectl, dig or nslookup does indeed work:
```
$ resolvectl query -t DS net.
net IN DS 35886 8 2 7862b27f5f516ebe19680444d4ce5e762981931842c465f00236401d8bd973ee -- link: VPN

-- Information acquired via protocol DNS in 2.8ms.
-- Data is authenticated: yes
```
```
$ dig DS net. @127.0.0.53

; <<>> DiG 9.16.7 <<>> DS net. @127.0.0.53
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 58646
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 65494
;; QUESTION SECTION:
;net.                           IN      DS

;; ANSWER SECTION:
net.                    6316    IN      DS      35886 8 2 7862B27F5F516EBE19680444D4CE5E762981931842C465F00236401D 8BD973EE

;; Query time: 0 msec
;; SERVER: 127.0.0.53#53(127.0.0.53)
;; WHEN: Sun Oct 04 18:33:53 CEST 2020
;; MSG SIZE  rcvd: 80

```
```
$ nslookup -querytype=DS net. 127.0.0.53
Server:         127.0.0.53
Address:        127.0.0.53#53

Non-authoritative answer:
net     rdata_43 = 35886 8 2 7862B27F5F516EBE19680444D4CE5E762981931842C465F00236401D 8BD973EE

Authoritative answers can be found from:

```

So it appears delv does something different here which lets resolved fail. 
Should this be reported to them or is it expected to be like this?

**Comment by mercora** [comment]
after some further investigation with tcpdump it appears only when asking delv to validate "net." directly resolved returns ServFail but not for the DS query but for an A query it does beforehand. That is probably expected then. trying to validate "lileth.net." with delv works better but resolved apparently filters rrsig responses which then make it fail to validate.

So it appears it only forbids RRSIG queries but filters them from the response.
Is that true?

**Comment by pemensik** [comment]
Because delv makes DNSSEC validation inside the tool. You can try ``delv +cd``, which would disable validation in forwarder. It would still process validation on your machine. And validation has to fail, when the resolver is stripping out rrsig records. I made a request in issue #17226, to implement always support for DNSSEC. And instead toggle just validation. There is no need for implementations done after DNSSEC is widespread to pretend it does not know nothing about it. It should just allow validation exceptions to support broken network configurations.

**Comment by mercora** [comment]
well once i saw delv would not get the rrsig in replies it was obvious why it is failing to validate itself... as far as i understand the cd flag would cause the resolver to pass invalid data even in the case validation fails. i don't really see how that would help if RRSIG records are stil missing. Anyhow, apparently resolved does not even support this and returns NotImp to this type of query. I guess delv expects the RRSIG to come alongside the query response as requested which just does not happen.


```
delv +cd -t SOA lileth.net
;; NOTIMP unexpected RCODE resolving 'lileth.net/SOA/IN': 127.0.0.53#53
;; resolution failed: SERVFAIL
```
it looks like this without the presence of the flag:
```
01:22:14.550055 IP 127.0.0.1.51902 > 127.0.0.53.53: 49092+ [1au] SOA? lileth.net. (51)

01:22:14.552212 IP 172.18.11.12.47267 > 172.18.11.1.53: 4853+% [1au] SOA? lileth.net. (62)
01:22:14.576042 IP 172.18.11.1.53 > 172.18.11.12.47267: 4853$ 2/0/1 SOA, RRSIG (225)
01:22:14.576234 IP 172.18.11.12.53560 > 172.18.11.1.53: 30731+% [1au] DNSKEY? lileth.net. (62)
01:22:14.597651 IP 172.18.11.1.53 > 172.18.11.12.53560: 30731 4/0/1 DNSKEY, DNSKEY, RRSIG, RRSIG (539)
01:22:14.597886 IP 172.18.11.12.43459 > 172.18.11.1.53: 15266+% [1au] DS? lileth.net. (62)
01:22:14.620900 IP 172.18.11.1.53 > 172.18.11.12.43459: 15266$ 2/0/1 DS, RRSIG (282)
01:22:14.621401 IP 172.18.11.12.46983 > 172.18.11.1.53: 47287+% [1au] DNSKEY? net. (55)
01:22:14.644324 IP 172.18.11.1.53 > 172.18.11.12.46983: 47287$ 3/0/1 DNSKEY, DNSKEY, RRSIG (775)
01:22:14.644822 IP 172.18.11.12.34176 > 172.18.11.1.53: 47401+% [1au] DS? net. (55)
01:22:14.667934 IP 172.18.11.1.53 > 172.18.11.12.34176: 47401$ 2/0/1 DS, RRSIG (367)
01:22:14.668344 IP 172.18.11.12.60733 > 172.18.11.1.53: 44714+% [1au] DNSKEY? . (51)
01:22:14.690366 IP 172.18.11.1.53 > 172.18.11.12.60733: 44714$ 4/0/1 DNSKEY, DNSKEY, DNSKEY, RRSIG (1139)

01:22:14.724529 IP 127.0.0.53.53 > 127.0.0.1.51902: 49092$ 1/0/1 SOA (110)
01:22:14.724914 IP 127.0.0.1.51902 > 127.0.0.53.53: 38465+ [1au] DS? net. (44)
01:22:14.725077 IP 127.0.0.53.53 > 127.0.0.1.51902: 38465$ 1/0/1 DS (103)
```

**Comment by pemensik** [comment]
Sure, you would need DNSSEC=yes or allow-downgrade to have at least something working. I addressed that in issue #17226. Instead of refusing DNSSEC requests, it should just set NTA on '.' (root). This issue is valid with or without DNSSEC support, it should work in both cases.

**Comment by mercora** [comment]
I am not quite sure i can still follow. However, i just want to say i have `DNSSEC=yes` set and resolved does indeed do DNSSEC validation just fine like seen on the capture. AFAICT delv expects to get the RRSIG records alongside the reply while the DO flag is set. resolved does validate the query successfully and also indicates this in the response but does not send the RRSIG records alongside. This appears to be a deliberate decision, so i am not sure if this is a bug or not. It probably depends on if this is allowed by the spec or if resolved should not indicate support for this or something.

**Comment by pemensik** [comment]
Oh. I am sorry I would not expected this. Indeed, even with ``DNSSEC=yes``, it just discards RRSIG. No matter what was requested. I though it would be issue #4621, but this is even more severe.

```
# resolvectl | grep DNSSEC
      DNSSEC setting: yes                 
    DNSSEC supported: yes                 
      DNSSEC setting: yes                          
    DNSSEC supported: yes                          
# dig +dnssec @127.0.0.53 | grep -i RRSIG
```

This should have its own bug. I think it is definitely bug, when it claims DNSSEC supported.

**Comment by poettering** [comment]
> If some crappy routers cannot handle rrsig query, it should be reported to their manufacturer. Do not fix bugs of others in your software. 

I am sorry. We have to work with what we live with. We cannot just freezes random people's hardware. Yes, totally, fix bugs where they are, if it's in the stack we can control, i.e. is free software. But for everything else it's pointless to insist the other folks did something wrong, because the net result is that things break. And fixing them then is not going to be easy: cheap vendors don't like to do firmware upgrades. and even if they did, do you honestly think regular people would know how to apply them?

I am sorry, but we have to care about what is actually deployed, and work with that.

**Comment by pemensik** [comment]
I think you are missing the point. Because you are blocking valid query on any deployment, because maybe some hardware may not like it, you made a bug in your own implementation.

Owner has to request that record manually, it is usually not produced by software on localhost. Manual only request should be fulfilled. Can you name model and vendor of device, where were this issue detected?

**Comment by Habbie** [comment]
> If some crappy routers cannot handle rrsig query, it should be reported to their manufacturer.

No. RRSIG queries do not make sense. Hence, #17272 is not a duplicate of this. #17272 is a very reasonable thing to ask (of a resolver, I'm not here to decide whether systemd-resolved fits in that box). RRSIG type queries are not reasonable ever, anywhere.

I suggest closing this in favour of the feature request in #17272.

**Comment by Rombobeorn** [comment]
Lennart Poettering wrote:
> We have to work with what we live with. We cannot just freezes random people's hardware.

That makes sense as an argument for why systemd-resolved should avoid
generating RRSIG requests of its own – which I understand it wouldn't
do anyway, so it's no problem.

The argument does not make sense as an argument for refusing requests
from clients. If a client sends a request that breaks a defective
server, and the request is valid according to the protocol
specification, then a middleman is not at fault for forwarding the
request.

If somebody has been living with such a defective wifi router and
without systemd-resolved, then obviously they're not sending RRSIG
queries, so there's no need for systemd-resolved to shield the broken
router from something that isn't happening anyway. If somebody else has
been sending RRSIG queries manually, and their home router doesn't have
that particular defect, then injecting systemd-resolved into their
/etc/resolv.conf will break their usecase for no reason.

While RRSIG queries may be of little use today, it's conceivable that
some future protocol extension might find a use for them. The extension
would be compatible with existing compliant implementations, given that
such queries were always valid, but noncompliant things like systemd-
resolved and those wifi routers would break it.

> Let's not forget that resolved is a tightly sandboxed
> DNS/mDNS/LLMNR/… stack that is supposed to be useful as a gatekeeper
> for name resolution requests: a client that itself is sandboxed
> should be able to use it to do reasonably safe lookups only, but not
> be able to send arbitrarily formatted packets. Now one could argue
> that RRSIG queries are not yet "arbitrarily formatted", but I think
> we should still filter them out.

So the idea is to prevent a malware-infected sandbox from mounting a
denial-of-service attack on the user's home router by requesting an
RRSIG record?

For such a gatekeeper to be able to keep malware contained in any
meaningful way, it would have to impose lots of restrictions on the
sandbox – restrictions that would break things left and right if
imposed on the whole system. It seems to me that a DNS gatekeeper and a
system-wide name resolution service should be two separate programs
with rather different feature sets.

**Comment by pemensik** [comment]
> I suggest closing this in favour of the feature request in #17272.
I admit this issue has much lower severity than the closed one. If I have to choose, close this one and reopen #17272. This is just nice to have, the closed is blocker to some functionality.

**Comment by Habbie** [comment]
This ticket still makes no sense, and the behaviour described (`Specified resource record type 46 may not be used in a query.`) is still correct.

**Comment by jacekmigacz** [comment]
I agree with @pemensik that working-around crappy network devices is not a good idea.

I don't agree with @pemensik that `systemd-resolved` should honor direct RRSIG RR requests.
The only real purpose I can think of is to allow some kind of client-side signature validation, but it is not systemd-resolved job to provide that.

Also, @mercora is reporting problem with fix proposed in #24133.

**Comment by Habbie** [comment]
> The only real purpose I can think of is to allow some kind of client-side signature validation, but it is not systemd-resolved job to provide that.

and if it was systemd-resolved's job, allowing RRSIG type queries would not actually help for that
